hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java

```diff
@@ -1,81 +1,122 @@
-    private ConfigConfiguration configConfiguration;
+    public static final Map<String, RemoteConnectedServiceConfig> EMPTY_CONNECTED_SERVICE_CONFIG_MAP = Collections.emptyMap();
+    public static final Map<String, RemoteMatchingServiceConfig> EMPTY_MATCHING_SERVICE_CONFIG_MAP = Collections.emptyMap();
+    public static final List<RemoteServiceProviderConfig> EMPTY_SERVICE_PROVIDER_CONFIG_LIST = Collections.emptyList();
+    public static final RemoteConfigCollection EMPTY_COLLECTION = new RemoteConfigCollection(null, null, EMPTY_CONNECTED_SERVICE_CONFIG_MAP, EMPTY_MATCHING_SERVICE_CONFIG_MAP, EMPTY_SERVICE_PROVIDER_CONFIG_LIST);
+    private SelfServiceConfig selfServiceConfig;
     private AmazonS3 s3Client;
+    private LoadingCache<String, RemoteConfigCollection> cache;
+    private ObjectMapper objectMapper;
 
     public S3ConfigSource(ConfigConfiguration configConfiguration, AmazonS3 s3Client) {
-        this.configConfiguration = configConfiguration;
-        this.s3Client = s3Client;
+        this.selfServiceConfig = configConfiguration.getSelfService();
+        if (selfServiceConfig != null && selfServiceConfig.isEnabled()){
+            this.s3Client = s3Client;
+            this.objectMapper = new ObjectMapper();
+            objectMapper.setDateFormat(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss"));
+
+            CacheLoader<String, RemoteConfigCollection> cacheLoader = new S3ConfigCacheLoader();
+            long expiry = selfServiceConfig.getCacheExpiryInSeconds();
+            TimeUnit unit = TimeUnit.SECONDS;
+            if (expiry == 0){
+                expiry = 1;
+                unit = TimeUnit.MILLISECONDS;
+            }
+            this.cache = CacheBuilder.newBuilder()
+                    .refreshAfterWrite(expiry, unit)
+                    .build(cacheLoader);
+        }
     }
 
-    /**
-     * TODO: Refactor this static method out with guice later
-     * @param configConfiguration
-     * @return
-     */
-    public static S3ConfigSource setupS3ConfigSource(ConfigConfiguration configConfiguration) {
-        AmazonS3 s3Client;
-        if(!configConfiguration.getSelfService().getS3AccessKeyId().isEmpty() &&
-                !configConfiguration.getSelfService().getS3SecretKeyId().isEmpty()) {
-            BasicAWSCredentials awsCreds = new BasicAWSCredentials(configConfiguration.getSelfService().getS3AccessKeyId(),
-                    configConfiguration.getSelfService().getS3SecretKeyId());
-            s3Client = AmazonS3ClientBuilder.standard()
-                    .withRegion(configConfiguration.getSelfService().getAwsRegion())
-                    .withCredentials(new AWSStaticCredentialsProvider(awsCreds))
-                    .build();
-        } else {
-            s3Client = AmazonS3ClientBuilder.defaultClient();
+    public RemoteConfigCollection getRemoteConfig(){
+        if ( selfServiceConfig == null || !selfServiceConfig.isEnabled()){
```

Comment:
Maybe selfServiceConfig should be optional?

Reply:
Thinking about it, we should maybe default it to be there, with enabled false.

Reply:
Yeah, I think that's possible with DW's configuration